ethereum/contracts/relayer/wormholeRelayer/WormholeRelayerGovernance.sol

```diff
@@ -0,0 +1,231 @@
+        executeUpgradeMigration();
+    }
+
+    function executeUpgradeMigration() internal virtual {
+        //override and implement in WormholeRelayer upon contract upgrade (if required)
+    }
+
+    function registerWormholeRelayerContract(bytes memory encodedVm) external {
+        (uint16 foreignChainId, bytes32 foreignAddress) =
+            parseAndCheckRegisterWormholeRelayerContractVm(encodedVm);
+
+        getRegisteredWormholeRelayersState().registeredWormholeRelayers[foreignChainId] =
+            foreignAddress;
+    }
+
+    function submitContractUpgrade(bytes memory encodedVm) external {
+        address newImplementation = parseAndCheckContractUpgradeVm(encodedVm);
+
+        _upgradeTo(newImplementation);
+
+        (bool success, bytes memory revertData) =
+            address(this).call(abi.encodeCall(this.checkAndExecuteUpgradeMigration, ()));
+
+        if (!success) {
+            revert ContractUpgradeFailed(revertData);
+        }
+    }
+
+    function setDefaultDeliveryProvider(bytes memory encodedVm) external {
+        address newProvider = parseAndCheckRegisterDefaultDeliveryProviderVm(encodedVm);
+
+        getDefaultDeliveryProviderState().defaultDeliveryProvider = newProvider;
+    }
+
+    // ------------------------------------------- PRIVATE -------------------------------------------
+    using BytesParsing for bytes;
+
+    function parseAndCheckRegisterWormholeRelayerContractVm(bytes memory encodedVm)
+        private
+        returns (uint16 foreignChainId, bytes32 foreignAddress)
+    {
+        bytes memory payload = verifyAndConsumeGovernanceVM(encodedVm);
+        uint256 offset = parseAndCheckPayloadHeader(
+            payload, GOVERNANCE_ACTION_REGISTER_WORMHOLE_RELAYER_CONTRACT, true
+        );
+
+        (foreignChainId, offset) = payload.asUint16Unchecked(offset);
+        (foreignAddress, offset) = payload.asBytes32Unchecked(offset);
+
+        checkLength(payload, offset);
+    }
+
+    function parseAndCheckContractUpgradeVm(bytes memory encodedVm)
+        private
+        returns (address newImplementation)
+    {
+        bytes memory payload = verifyAndConsumeGovernanceVM(encodedVm);
+        uint256 offset =
+            parseAndCheckPayloadHeader(payload, GOVERNANCE_ACTION_CONTRACT_UPGRADE, false);
+
+        bytes32 newImplementationWhFmt;
+        (newImplementationWhFmt, offset) = payload.asBytes32Unchecked(offset);
+        //fromWormholeFormat reverts if first 12 bytes aren't zero (i.e. if it's not an EVM address)
+        newImplementation = fromWormholeFormat(newImplementationWhFmt);
+
+        checkLength(payload, offset);
+    }
+
+    function parseAndCheckRegisterDefaultDeliveryProviderVm(bytes memory encodedVm)
+        private
+        returns (address newProvider)
+    {
+        bytes memory payload = verifyAndConsumeGovernanceVM(encodedVm);
+        uint256 offset =
+            parseAndCheckPayloadHeader(payload, GOVERNANCE_ACTION_UPDATE_DEFAULT_PROVIDER, true);
```

Comment:
the third parameter here ('allowUnset') should be false; otherwise, the chainId isn't checked and someone can register a different chain's default delivery provider on this chain